proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+class, `MatrixRefImpl` is used for the common code.
+
+Other functions in the API are implemented in terms of `MatrixRefImpl`. This
+allows the compiler to generate more useful diagnostics if wrong types are used.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_MUL_OPTIMAL, true> MatrixA =
+      {ROBuffer, /*offset=*/128, /*stride=*/0};
+
+  MatrixRef<DATA_TYPE_FLOAT16, 4, 4, MATRIX_LAYOUT_ROW_MAJOR, true>
+      MatrixB = {ROBuffer, /*offset=*/128, /*stride=*/16};
+
+  RWMatrixRef<DATA_TYPE_FLOAT16, 128, 256, MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL>
+      MatrixC = {RWBuffer, /*offset=*/64, /*stride=*/0};
+}
+```
+
+Template parameters:
+
+- `DT` - the type used to store elements of the matrix in memory
+- `M` - the 'M' dimension of the matrix
+- `K` - the 'K" dimension of the matrix
+- `ML` - the layout of the matrix in memory
+- `Transpose` - whether or not this matrix should be transposed before any
+  operations
+
+Members:
+
+- `Buffer` - the buffer that the matrix is stored in
+  - For `MatrixRef` this is a `ByteAddressBuffer`
+  - For `RWMatrixRef` this is a `RWByteAddresssBuffer`
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer where
+  the matrix is located.
+- `Stride` - the stride, in bytes, between rows/columns of the matrix. This
+  value is ignored if the matrix layout is `MATRIX_LAYOUT_MUL_OPTIMAL` or
+  `MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL`.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BufferTy, DataType DT, uint M, uint K, MatrixLayout ML,
+          bool Transpose>
+struct MatrixRefImpl {
+  BufferTy Buffer;
+  uint StartOffset;
+  uint Stride;
+};
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool Transpose = false>
+using MatrixRef = MatrixRefImpl<ByteAddressBuffer, DT, M, K, ML, Transpose>;
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool Transpose = false>
+using RWMatrixRef = MatrixRefImpl<RWByteAddressBuffer, DT, M, K, ML, Transpose>;
+
+} // namespace linalg
+} // namespace dx
+```
+
+### struct VectorRef
+
+`VectorRef` and `RWVectorRef` specify a reference to a vector in memory along
+with the format of each element.
+
+As with `MatrixRef`, two versions are provided - one for vectors stored in
+`ByteAddressBuffer` and another for `RWByteAddressBuffer`. A base class,
+`VectorRefImpl`, covers both of them.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  VectorRef<DATA_TYPE_FLOAT16> VectorA = {ROBuffer, /*offset=*/128};
+  VectorRef<DATA_TYPE_FLOAT32> VectorB = {ROBuffer, /*offset=*/128};
+  RWVectorRef<DATA_TYPE_SINT16> VectorC = {RWBuffer, /*offset=*/64};
+}
+```
+
+Template parameter:
+
+- `DT` - the data type of each element stored in the buffer
```

Comment:
- For `VectorRef` this is a `ByteAddressBuffer`
  - For `RWVectorRef` this is a `RWByteAddresssBuffer`

Reply:
?

Reply:
I think maybe you think that `DT` refers to the buffer type?

Maybe this would be easier to follow if the implementation came first.

Reply:
Misclick, meant to click the 'Buffer' line under members below.

Reply:
Ah, gotcha. Fixed.